packages/vega-parser/src/parsers/guides/guide-util.js

```diff
@@ -86,3 +87,80 @@ export function extendOffset(value, offset) {
     : !isObject(value) ? { value, offset }
     : { ...value, offset: extendOffset(value.offset, offset) };
 }
+
+const DISCRETE_SCALES = new Set(['ordinal', 'band', 'point']);
+
+function formatType(_, scaleType) {
+  const formatType = _('formatType');
+
+  if (formatType) {
+    return formatType;
+  }
+  if (scaleType === Time) {
+    return 'time'
+  }
+  if (scaleType === UTC) {
+    return 'utc'
+  }
+}
+
+function wrapFormat(text, format, formatType) {
+  if (format == null) {
+    return text
+  }
+  const func = formatType === "time" ? "timeFormat" : formatType === "utc" ? "utcFormat" : "format";
+  return `${func}(${text}, "${format}")`;
+}
+
+function domainText(scaleType, scaleName, format, formatType) {
+  return DISCRETE_SCALES.has(scaleType)
+    ? ` with values " + domain('${scaleName}')`
+    : ` from " + ${wrapFormat(
+        `domain("${scaleName}")[0]`,
+        format,
+        formatType
+      )} + " to " + ${wrapFormat(
+        `domain("${scaleName}")[1]`,
+        format,
+        formatType
+      )}`;
+}
+
+function titleText(_) {
+  let title = _('title')
+
+  if (isArray(title)) {
+    // only use the first row of multiline titles
+    title = title[0];
+  }
+
+  if (title) {
+    return isObject(title) ? ` showing " + ${title.signal} + "` : ` showing ${title}`;
```

Comment:
What if a non-signal object is passed in?

Reply:
This crashes and burns. Is it ever expected to happen, though? Should we default to printing `[Object]` instead of crashing?

Reply:
I'll add `isSignal(title)` to the condition.